### DISCUSSION TITLE
Manipulating the URL in the clicked link

### DISCUSSION THREAD

**Discussion by bozdevs** [swup/swup#959] [Help] [answered]
I want to manipulate the "href" attribute in a link with a hook.

```html
<a href="/about" data-preview>About</a>
```

When the above link is clicked, if that link has the "`data-preview`" attribute, I want swup to "`visit`" the link as "`current_url/?preview=target_url.path`".

Description:

```js
target_url: ".../about"
current_url: ".../example"
new_visit_url: ".../example?preview=about"
```

I used "`link:click`" and "`visit:start`" event hooks for this, but I couldn't figure out how to change the `URL`. Can you help?

**Comment by daun** [comment]
What have you tried inside those hooks? Setting `visit.to.url` should get you there.

**Reply by bozdevs** [reply]
@daun Thank you for your answer. Yes, I also reviewed the same place. But the relevant hooks also contain "`html`" data in that section.

![image](https://github.com/user-attachments/assets/f1d00667-d3f6-4891-8037-001b29ff6722)

I don't understand this situation. If I change the URL of a data that already contains HTML, wouldn't the result be the same?

**Reply by bozdevs** [reply]
Sorry, but the "`path`" I tested actually produces a "`404`" result, so the "history API" falls into a vicious circle that I can't make sense of. In other words, when we press back while in "`new_url`", it still stays on the same page. It's as if the history API is deleted. This problem made me think that there was an error somewhere, but it only occurs in paths that produce "`404`". You answered the question, but I couldn't understand what happens to the history API in "`404`" :D

**Reply by daun** [reply]
In `link:click` and `visit:start`, there should definitely not be any html, the new page hasn't loaded yet at this point.

**Reply by daun** [reply]
Swup doesn't specifically handle 404s, it treats them just like 200s and only handles 500s with special logic.

**Reply by bozdevs** [reply]
> In `link:click` and `visit:start`, there should definitely not be any html, the new page hasn't loaded yet at this point.

Probably because I used `console.log()`, and probably because "`promise`" was used in the swup project, it adds the html into the log afterwards. Otherwise, I couldn't make much sense of it either.

**Reply by bozdevs** [reply]
Finally now the functionality I wanted works, thank you for everything.

**Reply by daun** [reply]
Great. Please post the working solution here for others and your future self :)

**Comment by bozdevs** [comment, marked as answer]
## Manipulating the URL in the clicked link

1- For SEO, we add the permalink directly to the link.
```html
<a href="localhost/about">About</a>
```

2- To differentiate we add a `data-preview` attribute.
```html
<a href="localhost/about" data-preview>About</a>
```

3- By taking advantage of Swup's hooks, we actually create the connection we want.
```js
swup.hooks.on('link:click', (visit) => {
   const element = visit.trigger.el;
   if (document.contains(element)) {
      if (element.hasAttribute("data-preview") || element.closest("[data-preview]")) {
         var target_url = new URL(element.href),
                current_url = new URL(window.location.href);
         current_url.searchParams.set("preview", target_url.pathname);
         visit.to.url = current_url;
      }
   }
});
```
In this way, when the link is clicked, instead of visiting the "`localhost/about`" link, it visits the "`localhost/current_url?preview=/about`" link.

4- (Optional) Keeping scroll position, "if you are going to open the target in a modal like me"
```js
// if (document.contains(element)) {
// ....
   visit.scroll.reset = false;
// }
```
